src/tools/tools/serialize.fs

```diff
@@ -153,3 +164,519 @@ module internal Serialize =
             assert(false)
             Unchecked.defaultof<'T>
 
+    let inline memo (f: 'a -> 'b) =
+        let cache = new Dictionary<'a, 'b>()
```

Comment:
No, not thread safe. Serialization is inherently sequential, and making this thread safe would have a cost, both in performance and cognitive, which we wouldn't be using. That's why I'm keeping it in module Serialize, not Utils.